packages/server/utils/atlassian/jiraIssues.ts

```diff
@@ -0,0 +1,111 @@
+import getRedis from '../getRedis'
+import AtlassianManager, {
+  RateLimitError,
+  JiraIssueBean
+} from 'parabol-client/utils/AtlassianManager'
+import ms from 'ms'
+import jsonEqual from 'parabol-client/utils/jsonEqual'
+
+const ISSUE_TTL_MS = ms('2d')
+const DEFAULT_RETRY_AFTER_SEC = 5
+const MAX_RETRY_AFTER_SEC = 30
+const MAX_RETRIES = 4
+
+const applyBackoff = (retries, seconds) => Math.pow(2, retries) * seconds
+const addJitter = (seconds) => seconds + (Math.random() * (1.3 - 0.7) + 0.7)
+
+export const getIssue = async (
+  manager: AtlassianManager,
+  cloudId: string,
+  issueKey: string,
+  onSuccess: (
+    issue: JiraIssueBean<{description: string; summary: string}, {description: string}>
+  ) => void /* cb invoked when we do lazy async request */,
+  onError: (e: Error) => void /* error invoked when we do lazy async request */,
+  extraFieldIds: string[] = []
+) => {
+  const key = `jira:${cloudId}:${issueKey}:${JSON.stringify(extraFieldIds)}`
+  const redis = getRedis()
+  let cachedIssue = await redis.get(key)
```

Comment:
fetching the issue from atlassian should not have to wait for the result to come back from redis.

for your consideration:

```ts

interface CacheHit {
  resolve: undefined
  cachedIssue: string
}

interface CacheMiss {
  resolve: (value: unknown) => void
  cachedIssue: null
}

interface CacheInit {
  resolve: undefined
  cachedIssue: undefined
}

type ReturnValue = CacheInit | CacheMiss | CacheHit

export const getIssue = async (
  manager: AtlassianManager,
  cloudId: string,
  issueKey: string,
  pushUpdateToClient: (issue: JiraIssueRaw) => void /* cb invoked when we do lazy async request */,
  extraFieldIds: string[] = []
) => {
  const key = `jira:${cloudId}:${issueKey}:${JSON.stringify(extraFieldIds)}`
  const redis = getRedis()
  const returnValue = {resolve: undefined, cachedIssue: undefined} as ReturnValue
  const cachedIssuePromise = redis.get(key)
  const issueResPromise = manager.getIssue(cloudId, issueKey, extraFieldIds).then(async (res) => {
    await issueResPromise
    if (res instanceof RateLimitError) {
      // do rate limit handling things
      const {retryAt} = res
      const delay = Math.max(0, retryAt.getTime() - Date.now())
      if (delay > ms('10s')) return res
      await sleep(delay)
      return getIssue(manager, cloudId, issueKey, pushUpdateToClient, extraFieldIds)
    } else if (res instanceof Error) {
      // return an error
      return res
    }

    if (returnValue.cachedIssue === undefined) {
      // redis is guaranteed to have returned, this is a noop
    } else if (returnValue.cachedIssue === null) {
      // the cache was empty when atlassian returned
      const issueFromJira = stringify(res)
      await redis.set(key, issueFromJira, 'PX', ISSUE_TTL_MS)
      returnValue.resolve(res)
    } else {
      // there was a cached value in redis
      const issueFromJira = stringify(res)
      if (issueFromJira !== returnValue.cachedIssue) {
        // the value from jira is fresher than what we have in redis
        // update redis & push an update via pubsub
        redis.set(key, issueFromJira, 'PX', ISSUE_TTL_MS)
        pushUpdateToClient(res)
      }
      // if they're equal, do nothing
    }
  })

  const cachedIssue = await cachedIssuePromise
  if (cachedIssue) {
    returnValue.cachedIssue = cachedIssue
    return cachedIssue
  }
  return new Promise((resolve) => {
    // resolve only exists if there was no cached value
    returnValue.resolve = resolve
  })
}
```